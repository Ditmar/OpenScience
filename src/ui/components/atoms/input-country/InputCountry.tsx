import React, { useEffect, useState } from 'react';
import { MenuItem, Select, type SelectChangeEvent } from '@mui/material';
import type { IProps, Option } from './types/IProps';
import styles from './InputCountry.module.scss';
import { Icon } from '../../../utils/svg-icons/icons';
import globe from '../../../../assets/icons/globe.svg?raw';
import close_circle from '../../../../assets/icons/close_circle.svg?raw';
import info from '../../../../assets/icons/info.svg?raw';

export function InputCountry(props: IProps) {
  const { label, options, error, helperText, disabled, borderRadius, size } = props;
  const [country, setCountry] = React.useState('');
  const [keyboardFocus, setKeyboardFocus] = useState(false);
  const handleChange = (event: SelectChangeEvent) => {
    setCountry(event.target.value);
  };

  useEffect(() => {
    const handleKeyDown = () => {
      setKeyboardFocus(true);
    };
    const handleMouseDown = () => {
      setKeyboardFocus(false);
    };

    window.addEventListener('keydown', handleKeyDown);
    window.addEventListener('mousedown', handleMouseDown);

    return () => {
      window.removeEventListener('keydown', handleKeyDown);
      window.removeEventListener('mousedown', handleMouseDown);
    };
  }, []);

  return (
    <div className={styles['country-input']}>
      <div>
        <div>
          {globe && <Icon src={globe} />}
          <p
            className={`${disabled ? styles.disabled : ''} ${size ? styles[size] : ''}`}
          >
            {label ?? ''}
          </p>
        </div>
        <div>
          {close_circle && <Icon src={close_circle} className={disabled ? styles.disabled : ''} />}
        </div>
      </div>
      <div className={styles['select-wrapper']}>
        <Select
          id="country-select"
          label="country-select"
          value={country}
          onChange={handleChange}
          displayEmpty
          disabled={disabled}
          inputProps={{ 'aria-label': 'Without label' }}
          sx={{
            '& .MuiOutlinedInput-notchedOutline': {
              borderColor: disabled
                ? 'var(--color-neutral-gray-strong-50)'
                : error
                  ? 'var(--color-feedback-negative-500)'
                  : 'var(--color-neutral-gray-soft-100)',
              borderWidth: '1px',
            },
            '&:hover .MuiOutlinedInput-notchedOutline': {
              borderColor: disabled
                ? ''
                : error
                  ? 'var(--color-feedback-negative-500)'
                  : 'var(--color-brand-primary-500)',
            },
            '&.keyboard-focus .MuiOutlinedInput-notchedOutline': {
              borderColor: disabled
                ? 'var(--color-neutral-gray-strong-50)'
                : error
                  ? 'var(--color-feedback-negative-100)'
                  : 'var(--color-brand-primary-100)',
              borderWidth: disabled ? '1px' : '4px',
            },
            '&.mouse-focus.Mui-focused .MuiOutlinedInput-notchedOutline': {
              borderColor: error
                ? 'var(--color-feedback-negative-500)'
                : 'var(--color-brand-primary-500)',
              borderWidth: '1px',
            },
          }}
          className={`
                            ${styles['select-country']}
                            ${
                              borderRadius === 'circle'
                                ? styles['select-circle']
                                : borderRadius === 'semi'
                                  ? styles['select-semi']
                                  : ''
                            }
                            ${keyboardFocus ? 'keyboard-focus' : 'mouse-focus'}
                            ${disabled ? styles.disabled : ''}
                            ${size ? styles[size] : ''}
                            `}
          MenuProps={{
            MenuListProps: {
              sx: {
                '& .MuiMenuItem-root': {
                  transition: 'all 0.3s',
                  display: 'flex',
                  gap: '10px',
                  fontWeight: 300,
                  fontFamily: 'Poppins-Light, sans-serif',
                  fontSize: '1rem',
                  '&:focus': {
                    backgroundColor: 'var(--color-brand-primary-50)',
                    '&:active': {
                      backgroundColor: 'var(--color-brand-primary-50)',
                    },
                  },
                  '&:active': {
                    backgroundColor: 'var(--color-brand-primary-50)',
                  },
                },
              },
            },
          }}
        >
          {options.map((item: Option, index: number) => {
            const iconFlag = `./images/flags/${item.code}.svg`;
            const code = index > 0 ? item.code : '';
            return (
              <MenuItem value={code} key={code}>
                <img src={iconFlag} alt="" />
                {item.country}
              </MenuItem>
            );
          })}
        </Select>
      </div>
      <div>
        <div>
          {info && (
            <Icon
              data-testid="button-icon"
              src={info}
              className={disabled ? styles.disabled : error ? styles.error : ''}
            />
          )}
          <p
            className={`${disabled ? styles.disabled : error ? styles.error : ''} ${
              size ? styles[size] : ''
            }`}
          >
            {helperText}
          </p>
        </div>
        <div>{info && <Icon data-testid="button-icon" src={info} />}</div>
      </div>
    </div>
  );
}
