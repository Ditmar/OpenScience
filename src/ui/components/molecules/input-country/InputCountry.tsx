import React, { useEffect, useState } from 'react';
import { Box, MenuItem, Select, type SelectChangeEvent } from '@mui/material';
import type { IProps, Option } from './types/IProps';
import { InputSelect } from 'ui/components/atoms/input-select/InputSelect';
import { InputLabel } from 'ui/components/atoms/input-label/InputLabel';

const options: Option[] = [
  { country: 'Argentina', code: 'ar' },
  { country: 'Bolivia', code: 'bo' },
  { country: 'Chile', code: 'cl' },
  { country: 'PerÃº', code: 'pe' },
];

export function InputCountry(props: IProps) {
  const { label, options, error, helperText, disabled, borderRadius, size } = props;
  const [country, setCountry] = useState('');
  const [keyboardFocus, setKeyboardFocus] = useState(false);

  const handleChange = (event: SelectChangeEvent) => setCountry(event.target.value);

  useEffect(() => {
    const onKey = () => setKeyboardFocus(true);
    const onMouse = () => setKeyboardFocus(false);
    window.addEventListener('keydown', onKey);
    window.addEventListener('mousedown', onMouse);
    return () => {
      window.removeEventListener('keydown', onKey);
      window.removeEventListener('mousedown', onMouse);
    };
  }, []);

  return (
    <Box>
      <Box sx={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', mb: 1 }}>
        <InputLabel label='Select Country' />
      </Box>
      <Box sx={{ position: 'relative' }}>
        <InputSelect options={options} />
      </Box>
      <Box sx={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', mt: 1 }}>
        <InputLabel label='Hint Text' />
      </Box>
    </Box>
  );
}
