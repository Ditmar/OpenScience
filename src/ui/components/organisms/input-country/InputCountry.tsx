import { Box } from '@mui/material';
import { InputSelect } from '../../atoms/input-select/InputSelect';
import { InputLabel } from '../../molecules/input-label/InputLabel';
import type { IProps } from './types/IProps';

export function InputCountry(props: IProps) {
  const {
    label,
    leftUpperIcon,
    rightUpperIcon,
    leftBottomIcon,
    rightBottomIcon,
    options,
    error,
    helperText,
    disabled,
    borderRadius,
    size,
  } = props;

  return (
    <Box sx={{ width: 'max-content', display: 'flex', flexDirection: 'column', gap: '0.75rem' }}>
      <Box>
        <InputLabel
          label={label ?? ''}
          labelDisabled={disabled}
          labelSize={size}
          leftIcon={leftUpperIcon}
          rightIcon={rightUpperIcon}
          rightIconDisabled={disabled}
        />
      </Box>
      <Box sx={{ position: 'relative' }}>
        <InputSelect
          options={options}
          borderRadius={borderRadius}
          size={size}
          error={error}
          disabled={disabled}
        />
      </Box>
      <Box>
        <InputLabel
          label={helperText ?? ''}
          labelHelper
          labelError={error}
          labelSize={size}
          labelDisabled={disabled}
          leftIcon={leftBottomIcon}
          rightIcon={rightBottomIcon}
          leftIconError={error}
          leftIconDisabled={disabled}
        />
      </Box>
    </Box>
  );
}
