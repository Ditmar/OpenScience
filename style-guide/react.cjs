const { resolve } = require('node:path');
const project = resolve(process.cwd(), 'tsconfig.json');

module.exports = {
  env: {
    browser: true,
    es2021: true,
    es6: true,
    jest: true,
  },
  extends: [
    require.resolve('./rules/base'),
    require.resolve('./configurations/_base'),
    'plugin:react-hooks/recommended',
    'airbnb',
    'plugin:storybook/recommended',
    require.resolve('./configurations/typescript'),
    require.resolve('./configurations/vitest'),
  ],
  parserOptions: {
    project,
    babelOptions: {
      presets: ['@babel/preset-react'],
    },
  },
  plugins: ['react-refresh'],
  settings: {
    react: {
      version: 'detect',
    },
  },
  globals: {
    JSX: true,
  },
  rules: {
    'react/react-in-jsx-scope': 'off',
    'react/button-has-type': 'off',
    'react-refresh/only-export-components': [
      'warn',
      { allowExportNames: ['meta', 'links', 'headers', 'loader', 'action'] },
    ],
  },
};
