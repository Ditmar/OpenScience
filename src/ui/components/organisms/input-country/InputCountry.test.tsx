import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import type { ICountry } from 'ui/components/atoms/country-list/interfaces/Country.interface';
import { InputCountry } from './InputCountry';

const countryList: ICountry[] = [
  { code: 'ar', name: 'Argentina', dialCode: '+54' },
  { code: 'at', name: 'Austria', dialCode: '+43' },
  { code: 'bd', name: 'Bangladesh', dialCode: '+880' },
  { code: 'be', name: 'Belgium', dialCode: '+32' },
  { code: 'bo', name: 'Bolivia', dialCode: '+591' },
  { code: 'br', name: 'Brazil', dialCode: '+55' },
  { code: 'ca', name: 'Canada', dialCode: '+1' },
  { code: 'cl', name: 'Chile', dialCode: '+56' },
  { code: 'cn', name: 'China', dialCode: '+86' },
  { code: 'co', name: 'Colombia', dialCode: '+57' },
  { code: 'cr', name: 'Costa Rica', dialCode: '+506' },
  { code: 'cu', name: 'Cuba', dialCode: '+53' },
  { code: 'de', name: 'Germany', dialCode: '+49' },
  { code: 'dk', name: 'Denmark', dialCode: '+45' },
  { code: 'eg', name: 'Egypt', dialCode: '+20' },
  { code: 'pe', name: 'Peru', dialCode: '+51' },
];

describe('InputCountry Component', () => {
  it('renders with label', () => {
    render(<InputCountry label="Select Country" helperText="hint" countryList={countryList} />);
    expect(screen.getByText('Select Country')).toBeInTheDocument();
  });

  it('renders with helperText and error', () => {
    render(
      <InputCountry
        label="Select Country"
        countryList={countryList}
        error
        helperText="Error text"
      />,
    );
    expect(screen.getByText('Error text')).toBeInTheDocument();
  });

  it('renders disabled state', () => {
    render(
      <InputCountry
        label="Select Country"
        countryList={countryList}
        helperText="Hint Text"
        disabled
      />,
    );
    const select = screen.getByRole('combobox');
    expect(select).toHaveAttribute('aria-disabled', 'true');
  });

  it('changes value on select', async () => {
    render(<InputCountry countryList={countryList} />);

    const button = screen.getByRole('combobox');
    fireEvent.mouseDown(button);

    const listbox = await screen.findByRole('listbox');

    const option = Array.from(listbox.querySelectorAll('[role="option"]')).find(
      (el) => el.textContent?.includes('Argentina'),
    );

    expect(option).toBeDefined();

    if (option) {
      fireEvent.click(option);
    }

    await waitFor(() => {
      expect(screen.getByText(/Argentina/i)).toBeInTheDocument();
    });
  });

  it('matches snapshot in empty state', () => {
    const { container } = render(
      <InputCountry label="Country" countryList={countryList} helperText="Hint Text" />,
    );
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with error', () => {
    const { container } = render(
      <InputCountry label="Country" countryList={countryList} error helperText="Hint Text" />,
    );
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot disabled', () => {
    const { container } = render(
      <InputCountry label="Country" countryList={countryList} helperText="Hint Text" disabled />,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders with custom borderRadius', () => {
    render(<InputCountry label="Select Country" countryList={countryList} borderRadius="circle" />);
    const select = screen.getByRole('combobox');
    expect(select).toBeInTheDocument();
  });

  it('renders with custom colorVariant', () => {
    render(
      <InputCountry label="Select Country" countryList={countryList} colorVariant="primary" />,
    );
    const select = screen.getByRole('combobox');
    expect(select).toBeInTheDocument();
  });

  it('renders with small size', () => {
    render(<InputCountry label="Select Country" countryList={countryList} size="small" />);
    const select = screen.getByRole('combobox');
    expect(select).toBeInTheDocument();
  });

  it('renders with medium size', () => {
    render(<InputCountry label="Select Country" countryList={countryList} size="medium" />);
    const select = screen.getByRole('combobox');
    expect(select).toBeInTheDocument();
  });

  it('renders with custom borderRadiusMenu and matches snapshot', () => {
    const { container } = render(
      <InputCountry label="Select Country" countryList={countryList} borderRadiusMenu="radius" />,
    );
    expect(container).toMatchSnapshot();
  });
});
