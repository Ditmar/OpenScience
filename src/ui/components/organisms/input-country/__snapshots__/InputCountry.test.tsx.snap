// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputCountry Component > matches snapshot disabled 1`] = `
<div>
  <div
    class="MuiBox-root css-1x1e0l8"
  >
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiBox-root css-1lekzkb"
      >
        <div
          class="MuiBox-root css-19nzcwv"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-1b4s42d-MuiTypography-root"
          >
            Country
          </p>
        </div>
        <div
          class="MuiBox-root css-0"
        />
      </div>
    </div>
    <div
      class="MuiBox-root css-79elbk"
    >
      <div
        class="MuiBox-root css-79elbk"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-disabled mouse-focus MuiSelect-root css-1rw9qf-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        >
          <div
            aria-disabled="true"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-label="Without label"
            aria-labelledby="country-select"
            class="MuiSelect-select MuiSelect-outlined Mui-disabled MuiInputBase-input MuiOutlinedInput-input Mui-disabled css-w76bbz-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            id="country-select"
            role="combobox"
          >
            <img
              alt="Argentina"
              class="MuiBox-root css-19rsff"
              src="./images/flags/ar.svg"
            />
            Argentina
          </div>
          <input
            aria-hidden="true"
            aria-invalid="false"
            class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
            disabled=""
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined Mui-disabled css-lohd6h-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w4cd9x"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiBox-root css-1lekzkb"
      >
        <div
          class="MuiBox-root css-19nzcwv"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-1b4s42d-MuiTypography-root"
          >
            Hint Text
          </p>
        </div>
        <div
          class="MuiBox-root css-0"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`InputCountry Component > matches snapshot in empty state 1`] = `
<div>
  <div
    class="MuiBox-root css-1x1e0l8"
  >
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiBox-root css-1lekzkb"
      >
        <div
          class="MuiBox-root css-19nzcwv"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-eq936y-MuiTypography-root"
          >
            Country
          </p>
        </div>
        <div
          class="MuiBox-root css-0"
        />
      </div>
    </div>
    <div
      class="MuiBox-root css-79elbk"
    >
      <div
        class="MuiBox-root css-79elbk"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary mouse-focus MuiSelect-root css-qgf974-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-label="Without label"
            aria-labelledby="country-select"
            class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-w76bbz-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            id="country-select"
            role="combobox"
            tabindex="0"
          >
            <img
              alt="Argentina"
              class="MuiBox-root css-19rsff"
              src="./images/flags/ar.svg"
            />
            Argentina
          </div>
          <input
            aria-hidden="true"
            aria-invalid="false"
            class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-lohd6h-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w4cd9x"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiBox-root css-1lekzkb"
      >
        <div
          class="MuiBox-root css-19nzcwv"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-9lmazj-MuiTypography-root"
          >
            Hint Text
          </p>
        </div>
        <div
          class="MuiBox-root css-0"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`InputCountry Component > matches snapshot with error 1`] = `
<div>
  <div
    class="MuiBox-root css-1x1e0l8"
  >
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiBox-root css-1lekzkb"
      >
        <div
          class="MuiBox-root css-19nzcwv"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-eq936y-MuiTypography-root"
          >
            Country
          </p>
        </div>
        <div
          class="MuiBox-root css-0"
        />
      </div>
    </div>
    <div
      class="MuiBox-root css-79elbk"
    >
      <div
        class="MuiBox-root css-79elbk"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary mouse-focus MuiSelect-root css-c8un2q-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-label="Without label"
            aria-labelledby="country-select"
            class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-w76bbz-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            id="country-select"
            role="combobox"
            tabindex="0"
          >
            <img
              alt="Argentina"
              class="MuiBox-root css-19rsff"
              src="./images/flags/ar.svg"
            />
            Argentina
          </div>
          <input
            aria-hidden="true"
            aria-invalid="false"
            class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-lohd6h-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w4cd9x"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiBox-root css-1lekzkb"
      >
        <div
          class="MuiBox-root css-19nzcwv"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ek0aq9-MuiTypography-root"
          >
            Hint Text
          </p>
        </div>
        <div
          class="MuiBox-root css-0"
        />
      </div>
    </div>
  </div>
</div>
`;
